File: lib/ts/models/src/linked-art/LinkedArtProduction.ts
import { crm } from "@paradicms/vocabularies";
import { Mixin } from "ts-mixer";
import { Memoize } from "typescript-memoize";
import { Agent } from "../Agent";
import { EventDerivedDatesMixin } from "../EventDerivedDatesMixin";
import { Image } from "../Image";
import { PartialDateTimeDescription } from "../PartialDateTimeDescription";
import { ThumbnailSelector } from "../ThumbnailSelector";
import { WorkCreationEvent } from "../WorkCreationEvent";
import { LinkedArtModel } from "./LinkedArtModel";
import { LinkedArtPerson } from "./LinkedArtPerson";
import { LinkedArtTimeSpan } from "./LinkedArtTimeSpan";
import { mapTermToLinkedArtModel } from "./mapTermToLinkedArtModel";

export class LinkedArtProduction
  extends Mixin(LinkedArtModel, EventDerivedDatesMixin)
  implements WorkCreationEvent {
  get agents(): readonly Agent[] {
    return this.creators;
  }

  readonly contributors = [];

  get creators(): readonly Agent[] {
    return this.carriedOutBy;
  }

  @Memoize()
  get carriedOutBy(): readonly LinkedArtPerson[] {
    return this.filterAndMapObjects(crm.P14_carried_out_by, term =>
      mapTermToLinkedArtModel(this, term)
    ).filter(model => model instanceof LinkedArtPerson) as LinkedArtPerson[];
  }

  readonly date = null;
  readonly description = null;

  override get displayDate(): string | null {
    return this.hasTimeSpan?.displayDate ?? super.displayDate;
  }

  get endDate(): PartialDateTimeDescription | null {
    return this.hasTimeSpan?.endOfTheEnd ?? null;
  }

  @Memoize()
  get hasTimeSpan(): LinkedArtTimeSpan | null {
    return this.findAndMapObject(crm["P4_has_time-span"], term => {
      const model = mapTermToLinkedArtModel(this, term);
      return model instanceof LinkedArtTimeSpan
        ? (model as LinkedArtTimeSpan)
        : null;
    });
  }

  readonly images = [];

  override get label(): string {
    if (super.label) {
      return super.label;
    } else {
      return "Production";
    }
  }

  readonly location = null;

  override preMemoize(): void {
    super.preMemoize();
    this.preMemoizeEventDerivedDates();
    this.carriedOutBy;
  }

  get startDate(): PartialDateTimeDescription | null {
    return this.hasTimeSpan?.beginOfTheBegin ?? null;
  }

  thumbnail(selector: ThumbnailSelector): Image | null {
    return null;
  }

  readonly type: "WorkCreation" = "WorkCreation";
  readonly workLocation = null;
}

File: lib/py/etl/paradicms_etl/models/linked_art/linked_art_production.py
from rdflib import URIRef

from paradicms_etl.models.event import Event
from paradicms_etl.models.linked_art.linked_art_model import LinkedArtModel
from paradicms_etl.models.linked_art.linked_art_person import LinkedArtPerson
from paradicms_etl.models.linked_art.linked_art_time_span import LinkedArtTimeSpan
from paradicms_etl.namespaces.crm import CRM


class LinkedArtProduction(LinkedArtModel, Event):
    @property
    def carried_out_by(self) -> tuple[LinkedArtPerson, ...]:
        return tuple(
            value
            for value in self._values(
                CRM.P14_carried_out_by, self._map_term_to_linked_art_model
            )
            if isinstance(value, LinkedArtPerson)
        )

    @property
    def has_time_span(self) -> LinkedArtTimeSpan | None:
        time_span = self._optional_value(
            CRM["P4_has_time-span"], self._map_term_to_linked_art_model
        )
        return time_span if isinstance(time_span, LinkedArtTimeSpan) else None

    @classmethod
    def rdf_type_uri(cls) -> URIRef:
        return CRM.E12_Production
